### ISSUE TITLE
disconnecting from Server

### ISSUE THREAD

**Issue by TPCQitek** [sinricpro/esp8266-esp32-sdk#293]
Hello, can you please advise on this disconnect error?  thanks 
Background is the Example SWITCH is fine, remains on line, works fine.
however, when I blend this code to my main code, the server is disconnecting,

I understood the LIB WebSockets are current. but this looks related to WebSocket.

**Boot up message:**

16:22:30.816 -> ******************************************
16:22:30.816 -> [SinricPro:Websocket]: Connecting to WebSocket Server using SSL (ws.sinric.pro)
16:22:30.816 -> [SinricPro:Websocket]: headers: 
16:22:30.816 -> appkey:13984adb-f254-48d2-8b69-158fc6359094
16:22:30.816 -> deviceids:6349fcef16440f13ff7d36db
16:22:30.816 -> restoredevicestates:true
16:22:30.816 -> ip:192.168.0.35
16:22:30.816 -> mac:94:B9:7E:79:C5:68
16:22:30.816 -> platform:ESP32
16:22:30.816 -> SDKVersion:2.9.17

**operating LOG:**

16:22:37.215 -> [E][ssl_client.cpp:36] _handle_error(): [start_ssl_client():216]: (-15104) PK - The pubkey tag or value is invalid (only RSA and EC are supported)
16:22:37.215 -> [E][WiFiClientSecure.cpp:133] connect(): start_ssl_client: -15104
16:22:37.215 -> [SinricPro:Websocket]: disconnected
16:22:37.215 -> Disconnected from SinricPro
16:22:40.103 -> [E][ssl_client.cpp:36] _handle_error(): [start_ssl_client():216]: (-15104) PK - The pubkey tag or value is invalid (only RSA and EC are supported)
16:22:40.150 -> [E][WiFiClientSecure.cpp:133] connect(): start_ssl_client: -15104
16:22:40.150 -> [SinricPro:Websocket]: disconnected
16:22:40.150 -> Disconnected from SinricPro
16:22:43.004 -> [E][ssl_client.cpp:36] _handle_error(): [start_ssl_client():216]: (-10368) X509 - Allocation of memory failed
16:22:43.004 -> [E][WiFiClientSecure.cpp:133] connect(): start_ssl_client: -10368
16:22:43.004 -> [SinricPro:Websocket]: disconnected
16:22:43.004 -> Disconnected from SinricPro

Thanks Terence

**Comment by kakopappa** [comment]
X509 - Allocation of memory failed

Looks like running out of memory. If you are using a esp8266, try on ESP32

**Comment by TPCQitek** [comment]
Aruna   hello thanks for reply,

I am using the ESP32  8mB vroom module,

when I load the SWITCH example and credentials, its fine.

but integrated the example into the main code, (which is quite large) i get
this error.

see thi log from the IDE .

I notice the clock reporting 38.13 MHZ

**********************************************  IDE ***********************
Sketch uses 1893562 bytes (56%) of program storage space. Maximum is
3342336 bytes.
Global variables use 62660 bytes (19%) of dynamic memory, leaving 265020
bytes for local variables. Maximum is 327680 bytes.
C:\Users\user\Documents\ArduinoData\packages\esp32\tools\esptool_py\3.0.0/esptool.exe
--chip esp32 --port COM3 --baud 921600 --before default_reset --after
hard_reset write_flash -z --flash_mode dio --flash_freq 80m --flash_size
detect 0xe000
C:\Users\user\Documents\ArduinoData\packages\esp32\hardware\esp32\1.0.6/tools/partitions/boot_app0.bin
0x1000
C:\Users\user\Documents\ArduinoData\packages\esp32\hardware\esp32\1.0.6/tools/sdk/bin/bootloader_qio_80m.bin
0x10000
C:\Users\user\AppData\Local\Temp\arduino_build_478642/QTG_httpAT161022.ino.bin
0x8000
C:\Users\user\AppData\Local\Temp\arduino_build_478642/QTG_httpAT161022.ino.partitions.bin
esptool.py v3.0-dev
Serial port COM3
Connecting........_____....._____....._____.
Chip is ESP32-D0WD-V3 (revision 3)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding
Scheme None
WARNING: Detected crystal freq 38.13MHz is quite different to normalized
freq 40MHz. Unsupported crystal in use?
Crystal is 40MHz
MAC: 94:b9:7e:79:c5:68
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Auto-detected Flash size: 16MB
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.0 seconds (effective
4681.1 kbit/s)...
Hash of data verified.
Flash params set to 0x024f
Compressed 18656 bytes to 12052...
Writing at 0x00001000... (100 %)
Wrote 18656 bytes (12052 compressed) at 0x00001000 in 0.1 seconds
(effective 1036.4 kbit/s)...
Hash of data verified.
Compressed 1893680 bytes to 1119355...
Writing at 0x00010000... (1 %)
Writing at 0x00014000... (2 %)


Writing at 0x0011c000... (98 %)
Writing at 0x00120000... (100 %)
Wrote 1893680 bytes (1119355 compressed) at 0x00010000 in 18.1 seconds
(effective 837.2 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 129...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (129 compressed) at 0x00008000 in 0.0 seconds (effective
1638.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...







Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by sivar2311** [comment]
Hi @TPCQitek !

> I am using the ESP32  8mB vroom module,
> Global variables use 62660 bytes (19%) of dynamic memory, leaving 265020
> bytes for local variables. Maximum is 327680 bytes.

The module series "WROOM" has 4, 8 or 16 MB FLASH-ROM and **320 KB** RAM.
Your sketch leaves about 260KB free for dynamic memory allocation. As far as I know, SSL requires about 120 to 130 KB. 

Try your Sketch without SSL:
To do so, add this line to your sketch (before including SinricPro.h):
`#define SINRICPRO_NOSSL`

> WARNING: Detected crystal freq 38.13MHz is quite different to normalized

This could indeed lead to problems and should be checked.

**Comment by TPCQitek** [comment]
Thanks, I'll check it out.

yes Xtal is other issue, I need to mount up moer 8mB module to isolate
that,,  strange, (to me)


Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by TPCQitek** [comment]
Hello . ok I used the _NOSSL and its remaining on line, I will look at the
memory use and I can use a method to open that up,.

OK thanks for that.  good for now, its not urgent so I can do more testing,

also the "OK Google: please turn ON or OFF is fine and button is fine,


*so if I can also ask about the AirQuality example please?*

are there any "utterances" on Google that I can use to ask for "what's my
Air quality now?"   I can see the json file output, but how to utilize that

what is the CO2 ?  whats the Air Quality now?


I have sensors to Temp - Humidity, CO2, PM2.5, PM 10 , NO2 , CH20 and O3
and Also Air Quality rating,


thanks heaps,  :)



Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by sivar2311** [comment]
Referring to [Google's smart home device types](https://developers.google.com/assistant/smarthome/guides), the air quality sensor is not on the list.

@kakopappa Are you aware of Google Home's support for air quality sensors?

**Comment by kakopappa** [comment]
You have to use the sinric pro app for now. Current Google Home
implementation does not support Air Quality but we can plan for it in the
future via SensorState

https://developers.google.com/assistant/smarthome/traits/sensorstate

**Comment by TPCQitek** [comment]
ok thanks.. I will read more,

we have Air Quality management. The main market is commercial . we have
extensive new quality sensors, using the Dashboards with ThingsBoard
see ThingsBoard
Professional Edition Documentation | ThingsBoard Professional Edition
<https://thingsboard.io/docs/pe/>

 so the Google home is more a demonstration etc or some smaller product,
would be "nice to have" if we can ask about the air quality etc on the
voice.

Our Webpage is out of date needs a rebuild, but Covid got in the way, sorry
about that, due to do new release in November

Opportunities for Established Distributors – QiTek Group
<https://www.qitekgroup.com/distributors/>


thanks for your help, we can follow up on Air Quality..


Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by TPCQitek** [comment]
hello ok I will wait on the AirQuality update..

can you please explain the use of?

 SinricPro.restoreDeviceStates(true); // Uncomment to restore the last
known state from the server.

when is this possibly required?


Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by TPCQitek** [comment]
Ok tks .

**Comment by TPCQitek** [comment]
Hello, are you aware of a LIB / github example that can be used to set up
the Google Home via Bluetooth ESP32?

 thanks..


Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by TPCQitek** [comment]
Hello ,, any idea on release date? for Air Quality,

thanks


Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by kakopappa** [comment]
Today 25/Oct/2022 at 4.00 PM Thailand time we will do the production release.

**Comment by kakopappa** [comment]
The update is now available, you must relink the google home action to effect.

1. Hey Google, what's the PM2.5 / PM10 level in the bedroom (the room where your Air Quality Sensor is)

2. Google will automatically notify you when the air quality is unhealthy. (Must enable notifications in the Google Home app. Does not support assistant)

3. Google Home app does not show the PM2.5 | PM10 levels it seems. Can inquire via voice only.

DEMO: https://www.youtube.com/shorts/qaczcy4XaNE

**Comment by TPCQitek** [comment]
Ok ..I will have test..

Can I recommend/ suggestions of other features? Later

**Comment by TPCQitek** [comment]
Ok .

For interest here.
In my case I have existing Commands for Power on off . And management Lamps
and speeds .
So i treated the incoming msg as button press or as RPC schedule command.
My heart beat check sinric service and updates power state if the TOUCH lcd
of command changed it.

Works well

**Comment by TPCQitek** [comment]
Ok .

Not getting google sensible response .

I had / again restart registration iaq device.  Sinrinpro / HOME Google .

May be not rolled up yet to Google servers Australia?

**Comment by kakopappa** [comment]
It should be available everywhere. 

`Not getting google sensible response. `

Can you be more specific about the issue you are facing? does it show in Google Home? are sending the PM2.5 or PM10 values to the server and does it show in the dashboard? if it does, then it should work like in the demo.  For Google Home to notify you when the PM levels are unhealthy, you must enable the notifications from the Google Home app (Open the sensor) as well.  

Suggestions are welcome. We have a separate repo for feature-requests. https://github.com/sinricpro/feature-requests

**Comment by TPCQitek** [comment]
Yes got voice ok ..but not yet notification...
> Connected to SinricPro
15:15:03.897 -> Sending Air Quality event ..

remaining connected, the google home void responds with the PM2.5 and the
Air Quality,

Google Home Notifications are all ON,

but not giving the message, (IE like the Youtube demo)






I see the Notification in the debug, as send

Regards:

Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by kakopappa** [comment]
To see the google home notification, the sensor must report unhealthy pm2.5
level

**Comment by TPCQitek** [comment]
Hello, yes it was reporting in the debug console as json but the google
voice did not operate..


if I can ask please?

the APP_SECRET etc are these fixed size?
is there was to use smaller SECRET? as I am hitting some memory issues,
just looking at working around, but ask anyway. :(

Regards:


Terence Curtis

CONTACTS:
                       SKYPE: terencepcurtis

Mobile:            Australia   +61 412 262 646




************************************************************************************

**Comment by sivar2311** [comment]
> the APP_SECRET etc are these fixed size? is there was to use smaller SECRET? as I am hitting some memory issues, just looking at working around, but ask anyway. :(

Hi @TPCQitek !
APP_SECRET is a C-string with a length of 73 bytes + additional null terminating byte (74 bytes in total).
This should not cause any "memory issues".

You cannot change the length of the APP_SECRET! 
This is the key with which the messages to the SinricPro server are signed. 
If you change this key, your messages will no longer contain valid signatures and (for security reasons) will be rejected by the server as invalid.

**Comment by TPCQitek** [comment]
Yes ok understand. It was a question. I was looking at managing quite a
number of memory items . As I have several services running .

I think I have solved the problem..

Thanks .

**Comment by sivar2311** [comment]
The original issue was solved in https://github.com/sinricpro/esp8266-esp32-sdk/issues/293#issuecomment-1283398020 and is therefore closed.
For further questions or in case of problems, please feel free to open a [new issue](https://github.com/sinricpro/esp8266-esp32-sdk/issues).
